From f20cde033062241ae61227daeac08271e2ef4ba4 Mon Sep 17 00:00:00 2001
From: Chris Harrelson <chrishtr@chromium.org>
Date: Thu, 04 Jan 2024 00:07:19 +0000
Subject: [PATCH] Defer detach of the AX subtree until it's safe.

Tested by TabSearchUIBrowserTest.CloseTabSearchAsBrowserTabDoesNotCrash (crashed in force-a11y debug mode before this CL).

Ref: https://ci.chromium.org/ui/p/chromium/bisection/test-analysis/b/5637053519757312

Bug: 1512170

Change-Id: I358cf3f4699d745a4cc81646ada7e60ce488ec59
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5160323
Reviewed-by: Aaron Leventhal <aleventhal@chromium.org>
Commit-Queue: Chris Harrelson <chrishtr@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1242696}
---

--- a/third_party/blink/renderer/core/display_lock/display_lock_context.cc
+++ b/third_party/blink/renderer/core/display_lock/display_lock_context.cc
@@ -581,7 +581,7 @@
   // We also need to notify the AX cache (if it exists) to update the childrens
   // of |element_| in the AX cache.
   if (auto* ax_cache = element_->GetDocument().ExistingAXObjectCache()) {
-    ax_cache->SubtreeIsAttached(element_);
+    ax_cache->RemoveSubtreeWhenSafe(element_);
   }
 
   // Schedule ContentVisibilityAutoStateChange event if needed.
